File: composeApp/src/androidMain/kotlin/com/arcvgc/app/ui/battledetail/PokemonDetailCard.kt
package com.arcvgc.app.ui.battledetail

import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.OutlinedCard
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalInspectionMode
import org.jetbrains.compose.resources.painterResource
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import coil3.compose.AsyncImage
import coil3.compose.LocalPlatformContext
import coil3.request.ImageRequest
import coil3.request.crossfade
import com.arcvgc.app.shared.Res
import com.arcvgc.app.shared.preview_item
import com.arcvgc.app.shared.preview_tera
import com.arcvgc.app.ui.components.PokemonAvatar
import com.arcvgc.app.ui.components.PreviewAsyncImage
import com.arcvgc.app.ui.components.TypeIconRow
import com.arcvgc.app.ui.components.TypeInfo
import com.arcvgc.app.ui.model.ItemUiModel
import com.arcvgc.app.ui.model.PokemonDetailUiModel
import com.arcvgc.app.ui.model.TeraTypeUiModel
import com.arcvgc.app.ui.model.TypeUiModel
import com.arcvgc.app.ui.tokens.AppTokens.BulletSeparator
import com.arcvgc.app.ui.tokens.AppTokens.MoveChipCornerRadius

@Composable
fun PokemonDetailCard(
    pokemon: PokemonDetailUiModel,
    modifier: Modifier = Modifier,
    onPokemonClick: ((Int, String, String?, List<String>) -> Unit)? = null
) {
    OutlinedCard(
        modifier = modifier.fillMaxWidth(),
        colors = CardDefaults.outlinedCardColors(
            containerColor = MaterialTheme.colorScheme.surface
        ),
        border = CardDefaults.outlinedCardBorder(),
        onClick = if (onPokemonClick != null) {
            { onPokemonClick(pokemon.id, pokemon.name, pokemon.imageUrl, pokemon.types.mapNotNull { it.imageUrl }) }
        } else {
            {}
        }
    ) {
        val context = LocalPlatformContext.current
        val isPreview = LocalInspectionMode.current

        Box(modifier = Modifier.fillMaxWidth()) {
            // Main centered content
            Column(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(12.dp),
                horizontalAlignment = Alignment.CenterHorizontally
            ) {
                // Image area — pokemon overflows the circle background
                // Layer order: pokemon circle → pokemon image → item circle → item image
                Box(
                    modifier = Modifier.size(144.dp),
                    contentAlignment = Alignment.Center
                ) {
                    PokemonAvatar(
                        imageUrl = pokemon.imageUrl,
                        contentDescription = pokemon.name,
                        circleSize = 100.dp,
                        spriteSize = 144.dp
                    )

                    // Item circle background + image — over pokemon image
                    if (pokemon.item != null && isPreview) {
                        Box(
                            modifier = Modifier
                                .align(Alignment.BottomEnd)
                                .padding(4.dp)
                                .size(36.dp),
                            contentAlignment = Alignment.Center
                        ) {
                            Box(
                                modifier = Modifier
                                    .matchParentSize()
                                    .clip(CircleShape)
                                    .background(MaterialTheme.colorScheme.surfaceVariant)
                            )
                            Image(
                                painter = painterResource(Res.drawable.preview_item),
                                contentDescription = pokemon.item?.name,
                                modifier = Modifier.size(28.dp),
                                contentScale = ContentScale.Fit
                            )
                        }
                    } else {
                        pokemon.item?.imageUrl?.let { itemImageUrl ->
                            var itemImageLoaded by remember { mutableStateOf(false) }
                            Box(
                                modifier = Modifier
                                    .align(Alignment.BottomEnd)
                                    .padding(4.dp)
                                    .size(36.dp),
                                contentAlignment = Alignment.Center
                            ) {
                                if (itemImageLoaded) {
                                    Box(
                                        modifier = Modifier
                                            .matchParentSize()
                                            .clip(CircleShape)
                                            .background(MaterialTheme.colorScheme.surfaceVariant)
                                    )
                                }
                                AsyncImage(
                                    model = ImageRequest.Builder(context)
                                        .data(itemImageUrl)
                                        .crossfade(true)
                                        .build(),
                                    contentDescription = pokemon.item?.name,
                                    modifier = Modifier.size(28.dp),
                                    contentScale = ContentScale.Fit,
                                    onSuccess = { itemImageLoaded = true }
                                )
                            }
                        }
                    }
                }

                Spacer(modifier = Modifier.height(2.dp))

                // Pokemon name
                Text(
                    text = pokemon.name,
                    style = TextStyle(
                        fontWeight = FontWeight.Bold,
                        fontSize = 16.sp,
                    )
                )

                Spacer(modifier = Modifier.height(2.dp))

                // Ability · Item
                if (pokemon.abilityName != null || pokemon.item != null) {
                    Row(
                        verticalAlignment = Alignment.CenterVertically,
                        horizontalArrangement = Arrangement.Center
                    ) {
                        pokemon.abilityName?.let { ability ->
                            Text(
                                text = ability,
                                style = TextStyle(
                                    fontSize = 12.sp,
                                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                                )
                            )
                        }
                        pokemon.item?.let { item ->
                            if (pokemon.abilityName != null) {
                                Text(
                                    text = " $BulletSeparator ",
                                    style = TextStyle(
                                        fontSize = 12.sp,
                                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                                    )
                                )
                            }
                            Text(
                                text = item.name,
                                style = TextStyle(
                                    fontSize = 12.sp,
                                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                                )
                            )
                        }
                    }
                }

                // Moves — full-width rows with centered text
                val moves = pokemon.moves.take(4)
                if (moves.isNotEmpty()) {
                    Spacer(modifier = Modifier.height(6.dp))
                    Column(
                        modifier = Modifier.fillMaxWidth(),
                        verticalArrangement = Arrangement.spacedBy(2.dp)
                    ) {
                        Row(
                            modifier = Modifier.fillMaxWidth(),
                            horizontalArrangement = Arrangement.spacedBy(4.dp)
                        ) {
                            moves.getOrNull(0)?.let {
                                MoveChip(moveName = it, modifier = Modifier.weight(1f))
                            }
                            moves.getOrNull(1)?.let {
                                MoveChip(moveName = it, modifier = Modifier.weight(1f))
                            }
                        }
                        Row(
                            modifier = Modifier.fillMaxWidth(),
                            horizontalArrangement = Arrangement.spacedBy(4.dp)
                        ) {
                            moves.getOrNull(2)?.let {
                                MoveChip(moveName = it, modifier = Modifier.weight(1f))
                            }
                            moves.getOrNull(3)?.let {
                                MoveChip(moveName = it, modifier = Modifier.weight(1f))
                            }
                        }
                    }
                }
            }

            // Type icons — top start
            TypeIconRow(
                types = pokemon.types.map { TypeInfo(it.name, it.imageUrl) },
                modifier = Modifier
                    .align(Alignment.TopStart)
                    .padding(8.dp)
            )

            // Tera type icon — top end
            if (pokemon.teraType != null) {
                PreviewAsyncImage(
                    url = pokemon.teraType?.imageUrl,
                    previewDrawable = Res.drawable.preview_tera,
                    contentDescription = pokemon.teraType?.name,
                    modifier = Modifier
                        .align(Alignment.TopEnd)
                        .padding(8.dp)
                        .size(26.dp)
                )
            }
        }
    }
}

@Composable
private fun MoveChip(moveName: String, modifier: Modifier = Modifier) {
    Box(
        modifier = modifier
            .background(
                color = MaterialTheme.colorScheme.surfaceContainer,
                shape = RoundedCornerShape(MoveChipCornerRadius)
            )
            .padding(vertical = 2.dp),
        contentAlignment = Alignment.Center
    ) {
        Text(
            text = moveName,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
            fontSize = 11.sp
        )
    }
}

@Preview(showBackground = true)
@Composable
private fun PokemonDetailCardPreview() {
    MaterialTheme {
        PokemonDetailCard(
            pokemon = PokemonDetailUiModel(
                id = 149,
                name = "Dragonite",
                imageUrl = null,
                item = ItemUiModel(id = 1, name = "Choice Band", imageUrl = null),
                abilityName = "Multiscale",
                moves = listOf("Dragon Claw", "Extreme Speed", "Earthquake", "Ice Punch"),
                types = listOf(
                    TypeUiModel("Dragon", null),
                    TypeUiModel("Flying", null)
                ),
                teraType = TeraTypeUiModel(id = 1, name = "Normal", imageUrl = null)
            ),
            modifier = Modifier.padding(16.dp)
        )
    }
}

@Preview(showBackground = true)
@Composable
private fun PokemonDetailCardClosedTeamsheetPreview() {
    MaterialTheme {
        PokemonDetailCard(
            pokemon = PokemonDetailUiModel(
                id = 149,
                name = "Dragonite",
                imageUrl = null,
                item = null,
                abilityName = null,
                moves = emptyList(),
                types = listOf(
                    TypeUiModel("Dragon", null),
                    TypeUiModel("Flying", null)
                ),
                teraType = null
            ),
            modifier = Modifier.padding(16.dp)
        )
    }
}

@Preview(showBackground = true)
@Composable
private fun MoveChipPreview() {
    MaterialTheme {
        MoveChip(moveName = "Extreme Speed")
    }
}

File: iosApp/iosApp/PokemonDetailCard.swift
import SwiftUI
import Shared

struct PokemonDetailCard: View {
    let pokemon: PokemonDetailUiModel
    var onPokemonClick: ((Int32, String, String?, [String]) -> Void)? = nil

    var body: some View {
        Button {
            onPokemonClick?(pokemon.id, pokemon.name, pokemon.imageUrl, pokemon.types.compactMap { $0.imageUrl })
        } label: {
        ZStack(alignment: .topLeading) {
            // Main centered content
            VStack(spacing: 0) {
                // Image area — pokemon overflows the circle background
                ZStack {
                    PokemonAvatar(
                        imageUrl: pokemon.imageUrl,
                        circleSize: 100,
                        spriteSize: 144
                    )

                    // Item circle background + image — over pokemon image
                    if pokemon.item != nil && isPreview {
                        ZStack {
                            Circle()
                                .fill(Color(.systemGray5))
                            Image("PreviewItem")
                                .resizable()
                                .aspectRatio(contentMode: .fit)
                                .frame(width: 28, height: 28)
                        }
                        .frame(width: 36, height: 36)
                        .offset(x: 44, y: 44)
                    } else if let item = pokemon.item, let itemImageUrl = item.imageUrl {
                        ItemOverlay(imageUrl: itemImageUrl, name: item.name ?? "")
                            .frame(width: 36, height: 36)
                            .offset(x: 44, y: 44)
                    }
                }
                .frame(width: 144, height: 144)

                Spacer().frame(height: 2)

                // Pokemon name
                Text(pokemon.name)
                    .font(.system(size: 16, weight: .bold))
                    .lineLimit(1)

                Spacer().frame(height: 2)

                // Ability · Item
                if pokemon.abilityName != nil || pokemon.item != nil {
                    HStack(spacing: 0) {
                        if let abilityName = pokemon.abilityName {
                            Text(abilityName)
                                .font(.system(size: 12))
                                .foregroundColor(.secondary)
                        }
                        if let item = pokemon.item {
                            if pokemon.abilityName != nil {
                                Text(" \(AppTokens.bulletSeparator) ")
                                    .font(.system(size: 12))
                                    .foregroundColor(.secondary)
                            }
                            Text(item.name)
                                .font(.system(size: 12))
                                .foregroundColor(.secondary)
                        }
                    }
                    .lineLimit(1)
                }

                // Moves — full-width rows with centered text
                let moves = Array(pokemon.moves.prefix(4))
                if !moves.isEmpty {
                    Spacer().frame(height: 6)
                    VStack(spacing: 2) {
                        HStack(spacing: 4) {
                            if moves.count > 0 {
                                MoveChip(moveName: moves[0])
                            }
                            if moves.count > 1 {
                                MoveChip(moveName: moves[1])
                            }
                        }
                        HStack(spacing: 4) {
                            if moves.count > 2 {
                                MoveChip(moveName: moves[2])
                            }
                            if moves.count > 3 {
                                MoveChip(moveName: moves[3])
                            }
                        }
                    }
                }
            }
            .padding(12)
            .frame(maxWidth: .infinity)

            // Type icons — top start
            TypeIconRow(types: pokemon.types.map { (name: $0.name, imageUrl: $0.imageUrl) })
                .padding(8)

            // Tera type icon — top end
            if pokemon.teraType != nil {
                HStack {
                    Spacer()
                    PreviewAsyncImage(url: pokemon.teraType?.imageUrl, previewAsset: "PreviewTera")
                        .frame(width: 26, height: 26)
                        .padding(8)
                }
            }
        }
        .background(Color(.systemBackground))
        .cornerRadius(8)
        .overlay(
            RoundedRectangle(cornerRadius: 8)
                .stroke(Color(.opaqueSeparator), lineWidth: AppTokens.standardBorderWidth)
        )
        }
        .buttonStyle(PressableButtonStyle())
    }
}

private struct ItemOverlay: View {
    let imageUrl: String
    let name: String
    @State private var imageLoaded = false

    var body: some View {
        ZStack {
            if imageLoaded {
                Circle()
                    .fill(Color(.systemGray5))
            }
            AsyncImage(url: URL(string: imageUrl)) { phase in
                switch phase {
                case .success(let image):
                    image
                        .resizable()
                        .aspectRatio(contentMode: .fit)
                        .transition(.opacity.animation(.easeIn(duration: 0.2)))
                        .onAppear { imageLoaded = true }
                default:
                    Color.clear
                }
            }
            .frame(width: 28, height: 28)
        }
    }
}

struct MoveChip: View {
    let moveName: String

    var body: some View {
        Text(moveName)
            .font(.system(size: 11))
            .foregroundColor(Color(.label))
            .lineLimit(1)
            .frame(maxWidth: .infinity)
            .padding(.vertical, 6)
            .background(Color(.secondarySystemBackground))
            .cornerRadius(AppTokens.moveChipCornerRadius)
    }
}

#Preview {
    let samplePokemon = PokemonDetailUiModel(
        id: 149,
        name: "Dragonite",
        imageUrl: nil,
        item: ItemUiModel(id: 1, name: "Choice Band", imageUrl: nil),
        abilityName: "Multiscale",
        moves: ["Dragon Claw", "Extreme Speed", "Earthquake", "Ice Punch"],
        types: [
            TypeUiModel(name: "Dragon", imageUrl: nil),
            TypeUiModel(name: "Flying", imageUrl: nil)
        ],
        teraType: TeraTypeUiModel(id: 1, name: "Normal", imageUrl: nil)
    )

    PokemonDetailCard(pokemon: samplePokemon)
        .padding()
}

#Preview("Closed Teamsheet") {
    let closedPokemon = PokemonDetailUiModel(
        id: 149,
        name: "Dragonite",
        imageUrl: nil,
        item: nil,
        abilityName: nil,
        moves: [],
        types: [
            TypeUiModel(name: "Dragon", imageUrl: nil),
            TypeUiModel(name: "Flying", imageUrl: nil)
        ],
        teraType: nil
    )

    PokemonDetailCard(pokemon: closedPokemon)
        .padding()
}
